File: TicTacToe/NodeJS/MultiplayerGame.js
const readline = require('readline');

const Board = require('./Board');
const { gamePiece, gameStatus } = require('./enums');

const { X, O } = gamePiece;
const { WIN_O, WIN_X, TIE } = gameStatus

class MultiplayerGame {
    constructor() {
        this.board = new Board();
        this.player1 = {};
        this.player2 = {};
    }

    _askForMove(question, callback, low = 1, high = 9) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            prompt: `${this.currentTurn.name} ${question} (${low}-${high}): `
        });
        rl.on('line', line => {
            const number = Number(line.trim());
            if (number <= high && number >= low) {
                this.userAnswer = number;
                rl.close();
            } else
                rl.prompt();
        }).on('SIGINT', () => {
            console.log('\nThank you for playing TicTacToe. Have a good day ahead :)\n');
            process.exit(0);
        }).on('close', () => {
            callback(this.userAnswer);
        });
        rl.prompt();
    }

    _askName(prompt, onLine, onClose) {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            prompt: `${prompt}: `
        });
        rl.on('line', line => {
            if (line.trim()) {
                onLine(line);
                rl.close();
            } else 
                rl.prompt();
        }).on('SIGINT', () => {
            console.log('\nThank you for playing TicTacToe. Have a good day ahead :)\n');
            process.exit(0);
        }).on('close', () => {
            onClose();
        });
        rl.prompt();
    }

    decideFirstMove(next) {
        this._askName('Enter the name of player1', line => {
            this.player1.name = line.trim();
            this.player1.piece = X;
            this.currentTurn = this.player1;
        },
        () => {
            this._askName('Enter the name of player2', line => {
                this.player2.name = line.trim();
                this.player2.piece = O;
            },
            () => {
                next();
            });
        });
    }

    playerMove(next) {
        this._askForMove('where will you move?', res => {
            if (!this.board.isMoveLegal(res)) {
                console.log("This place is already occupied!");
                this.playerMove(next);
            } else {
                console.log('fine...\n');
                this.board.makeMove(res, this.currentTurn.piece);
                next();
            }
        });
    }

    printEnd(next) {
        if (this.status == TIE)
            console.log('It\'s a tie.\n');
        else if (this.status == WIN_O && this.player1.piece == O ||
            this.status == WIN_X && this.player1.piece == X)
            console.log(`${this.player1.name} Won!\n`);
        else
            console.log(`${this.player2.name} Won!\n`);
        next();
    }
}

module.exports = MultiplayerGame;

File: TicTacToe/C++/MultiplayerGame.cpp
#include "TicTacToe.hpp"
#include <iostream>
#include <string>

static std::string askName(const char * const req) {
    std::string name;
    std::cout << req << " : ";
    std::cin >> name;
    if (std::cin.fail()) {
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
    return name;
}

void TicTacToe::MultiplayerGame::decideFirstMove() {
    playerOne.name = askName("Enter player's one name");
    playerTwo.name = askName("Enter player's two name");
    playerOne.piece = X;
    playerTwo.piece = O;
}

void TicTacToe::MultiplayerGame::playerMove(std::string playerName) {
    std::string question = playerName + " where will you move?";
    int move;
    do {
        move = askNumber(question.c_str());
        if (!board.isMoveLegal(move))
            std::cout << "This place is already occupied!";
    }
    while (!board.isMoveLegal(move));
    std::cout << "fine...\n";
    if (playerOne.name == playerName)
        board.makeMove(move, playerOne.piece);
    else
        board.makeMove(move, playerTwo.piece);
}

void TicTacToe::MultiplayerGame::printEnd(GameStatus status) const {
    if (status == TIE)
        std::cout << "It's a tie.\n";
    else if (status == WIN_O)
        std::cout << playerTwo.name << " Won!\n";
    else
        std::cout << playerOne.name << " Won!\n";
}
